services/debugger/cache/embeddedcache/embeddedCache.go

```diff
@@ -0,0 +1,192 @@
+	config.RegisterDurationConfigVariable(15, &e.queryTimeout, false, time.Second, "LiveEvent.cache.queryTimeout")
+	config.RegisterIntConfigVariable(3, &e.retries, false, 1, "LiveEvent.cache.retries")
+}
+
+/*
+EmbeddedCache is an in-memory cache. Each key-value pair stored in this cache have a TTL and one goroutine removes the
+key-value pair form the cache which is older than TTL time.
+*/
+type EmbeddedCache[E any] struct {
+	Limiter      int
+	retries      int
+	path         string
+	Origin       string
+	done         chan struct{}
+	closed       chan struct{}
+	Ticker       time.Duration
+	queryTimeout time.Duration
+	CleanupFreq  time.Duration // TTL time on badgerDB
+	once         sync.Once
+	Db           *badger.DB
+	Logger       logger.Logger
+}
+
+type badgerLogger struct {
+	logger.Logger
+}
+
+func (badgerLogger) Errorf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) {
+	e.Init()
+	txn := e.Db.NewTransaction(true)
+	res := e.Read(key)
+	if len(res) >= e.Limiter {
+		res = res[len(res)-e.Limiter+1:]
+	}
+	res = append(res, value)
+	data, err := json.Marshal(res)
+	if err != nil {
+		return
+	}
+	setBadgerEntry := func(ctx context.Context) error {
+		entry := badger.NewEntry([]byte(key), data).WithTTL(e.CleanupFreq)
+		if err := txn.SetEntry(entry); err == badger.ErrTxnTooBig {
+			err = txn.Commit()
+			if err != nil {
+				return err
+			}
+			txn = e.Db.NewTransaction(true)
+			err = txn.SetEntry(entry)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	_ = misc.RetryWith(context.TODO(), e.queryTimeout, e.retries, setBadgerEntry)
+	_ = txn.Commit()
+}
+
+// Read fetches all the entries for a given key from badgerDB
+func (e *EmbeddedCache[E]) Read(key string) []E {
+	e.Init()
+	read := func(ctx context.Context) ([]E, error) {
+		var values []E
+		err := e.Db.View(func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+			prefix := []byte(key)
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				item := it.Item()
+				err := item.Value(func(v []byte) error {
+					err := json.Unmarshal(v, &values)
+					if err != nil {
+						return err
+					}
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+		return values, nil
+	}
+
+	res, _ := misc.QueryWithRetries(context.TODO(), e.queryTimeout, e.retries, read)
+
+	return res
+}
+
+func (e *EmbeddedCache[E]) Init() {
+	e.once.Do(func() {
+		e.loadCacheConfig()
+		badgerPathName := e.Origin + "/cache/badgerdbv3"
+		tmpDirPath, err := misc.CreateTMPDIR()
+		if err != nil {
+			e.Logger.Errorf("Unable to create tmp directory: %v", err)
+			return
```

Comment:
If we use a proper constructor we can fail there instead of logging and returning here. It'd be a more robust way to handle these cases in my opinion.